docs/design/persistent-volume-provisioning.md

```diff
@@ -0,0 +1,129 @@
+# Persistent Volume Provisioning
+
+This document proposes a model for dynamically provisioning [Persistent Volumes](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/persistent-storage.md)
+
+### tl;dr
+
+One new API kinds:
+
+A `PersistentVolumeController` (PVC) is a storage resource provisioned by an administrator.  PVCs do not have a namespace.  Like a ReplicationController maintains a minimum number of "replicas" of a pod in the system, so a PersistentVolumeController maintains a minimum number of "replicas" of a certain type/size of PersistentVolume available in the system.
+
+One new system component:
+
+`PersistentVolumeControllerManager` is a singleton control loop running in master that manages all PVControllers in the system.  The PVCM reconciles the current supply of available PersistentVolumes in the system with the desired levels according to the PVControllers.   This process is similar to the ```ReplicationManager``` that manages ReplicationControllers.
+
+Two new persistent volume functions:
+
+The PersistentVolumeSource interface requires ```Create``` and ```Recycle``` functions to support the dynamic creation and reclamation of persistent volumes.  Each storage provider implements its own PersistentVolumeSource interface.
```

Comment:
That is true.  If all claims got deleted, the number of available volumes might greatly exceed the "MinimumAvailable" count.  The pseudo-code for the controller contains a clause for recycling the excess volumes.  I suppose the difficulty lies in knowing what "recycle" means for some volumes.

For cloud volumes, deleting a volume for "recycling" seems like the best option.  That would decrease the available count.

For NFS, erasing a volume means it's still available and the loop that recycles the extras will loop forever and none get removed.

Let me think about this.  The plugin should probably advise the controller about what to do, since their "recycling" behavior is implementation specific.